src/utils/UrlUtils.ts

```diff
@@ -0,0 +1,182 @@
+        endpointParameters = { ...endpointParameters, url: path };
+      }
+
+      if (Utils.isNonEmptyArray(queryNormalized)) {
+        if (Utils.isNonEmptyArray(endpointParameters.queryString)) {
+          endpointParameters = {
+            ...endpointParameters,
+            queryString: Utils.concatWithoutDuplicate(endpointParameters.queryString, queryNormalized)
+          };
+        } else {
+          endpointParameters = { ...endpointParameters, queryString: queryNormalized };
+        }
+      }
+    });
+    return endpointParameters;
+  }
+
+  public static normalizeAsString(toNormalize: IUrlNormalize): string {
+    const { queryNormalized, path } = this.normalizeAsParts(toNormalize);
+
+    return `${path}${this.addToUrlIfNotEmpty(queryNormalized, '&', '?')}`;
+  }
+
+  public static normalizeAsParts(toNormalize: IUrlNormalize): IUrlNormalizedParts {
+    const pathsNormalized = this.normalizePaths(toNormalize);
+    const queryNormalized = this.normalizeQueryString(toNormalize);
+
+    return {
+      pathsNormalized,
+      queryNormalized,
+      path: this.addToUrlIfNotEmpty(pathsNormalized, '/', '')
+    };
+  }
+
+  private static normalizePaths(toNormalize: IUrlNormalize) {
+    return this.toArray(toNormalize.paths).map(path => {
+      if (Utils.isNonEmptyString(path)) {
+        return this.removeProblematicChars(path);
+      }
+      return '';
+    });
+  }
+
+  private static normalizeQueryString(toNormalize: IUrlNormalize) {
+    let queryNormalized: string[] = [];
+
+    if (toNormalize.queryAsString) {
+      const cleanedUp = this.toArray(toNormalize.queryAsString).map(query => {
+        query = this.removeProblematicChars(query);
+        query = this.encodeKeyValuePair(query);
+        return query;
+      });
+      queryNormalized = queryNormalized.concat(cleanedUp);
+    }
+
+    if (toNormalize.query) {
+      const paired: string[][] = pairs(toNormalize.query);
+      const mapped = paired.map(pair => {
+        const [key, value] = pair;
+
+        if (Utils.isNullOrUndefined(value) || Utils.isNullOrUndefined(key)) {
+          return '';
+        }
+
+        if (!this.isEncoded(value)) {
+          return [this.removeProblematicChars(key), Utils.safeEncodeURIComponent(value)].join('=');
+        } else {
+          return [this.removeProblematicChars(key), value].join('=');
+        }
+      });
+      queryNormalized = queryNormalized.concat(mapped);
+    }
+
+    return uniq(queryNormalized);
+  }
+
+  private static addToUrlIfNotEmpty(toAdd: string[], joinWith: string, leadWith: string) {
+    if (Utils.isNonEmptyArray(toAdd)) {
+      return `${leadWith}${compact(toAdd).join(joinWith)}`;
+    }
+    return '';
+  }
+
+  private static startsWith(searchString: string, targetString: string) {
+    return targetString.charAt(0) == searchString;
+  }
+
+  private static endsWith(searchString: string, targetString: string) {
+    return targetString.charAt(targetString.length - 1) == searchString;
```

Comment:
I wonder if there could be strange behaviour arising from the interplay of this method and `removeAtEnd` since this method compares only the last char of the target string, but `removeAtEnd` wants to slice more than just the last char. This could prevent the slice from taking place when the 'searchString` has length > 1.

Perhaps using a regex would be more aligned with the method name 'endsWith' whereas now the logic behaves like 'lastCharIs'.

```
return `/${searchString}$/`.test(targetString);
```
  
A similar argument applies to the `startsWith` method.

Edit: I saw that these methods are used only by `removeProblematicChars` which loops through an array of chars, so there would be no issues as is. I still believe the logic should be modified though, be it in `endsWith` or `removeFromEnd`.

Reply:
I'll go with dario's proposition and instead use a polyfill.